File: tests/typescript/test_add_two_nums.ts
// Basic harness for AddTwoNums (CommonJS-friendly)
const assert = require("node:assert").strict;
const path = require("node:path");

const rootDir = path.resolve(__dirname, "../..");
const modulePath = path.resolve(rootDir, "problems/typescript/AddTwoNums.ts");
const loaded = require(modulePath);

function resolveAddTwoNumbers(mod) {
  if (typeof mod === "function") {
    return mod;
  }

  if (mod && typeof mod.addTwoNumbers === "function") {
    return mod.addTwoNumbers.bind(mod);
  }

  if (mod && typeof mod.default === "function") {
    if (mod.default.prototype && typeof mod.default.prototype.addTwoNumbers === "function") {
      const solver = new mod.default();
      return solver.addTwoNumbers.bind(solver);
    }
    return mod.default;
  }

  if (mod && mod.default && typeof mod.default.addTwoNumbers === "function") {
    return mod.default.addTwoNumbers.bind(mod.default);
  }

  if (mod && typeof mod.Solution === "function") {
    const solver = new mod.Solution();
    if (typeof solver.addTwoNumbers === "function") {
      return solver.addTwoNumbers.bind(solver);
    }
  }

  throw new Error(
    "Could not find addTwoNumbers export. Export a function or a class with addTwoNumbers()."
  );
}

const ListNodeCtor = loaded.ListNode;

const addTwoNumbers = resolveAddTwoNumbers(loaded);

function listFromArray(values) {
  let head = null;
  let tail = null;

  for (const value of values) {
    const node = new ListNodeCtor(value, null);

    if (!head) {
      head = node;
      tail = node;
      continue;
    }

    tail.next = node;
    tail = node;
  }

  return head;
}

function arrayFromList(head) {
  const out = [];
  let current = head;

  while (current) {
    out.push(current.val);
    current = current.next;
  }

  return out;
}

function assertSum(l1, l2, expected) {
  const result = addTwoNumbers(listFromArray(l1), listFromArray(l2));
  assert.deepEqual(arrayFromList(result), expected);
}

assertSum([2, 4, 3], [5, 6, 4], [7, 0, 8]);
assertSum([0], [0], [0]);
assertSum([9, 9, 9, 9, 9, 9, 9], [9, 9, 9, 9], [8, 9, 9, 9, 0, 0, 0, 1]);

console.log("add two nums tests OK");

File: tests/python/test_add_two_nums.py
import sys
import unittest
from pathlib import Path

ROOT_DIR = Path(__file__).resolve().parents[2]
sys.path.append(str(ROOT_DIR / "problems" / "python"))

from AddTwoNums import ListNode, Solution


def list_from_array(values):
    head = None
    tail = None
    for value in values:
        node = ListNode(value)
        if head is None:
            head = node
            tail = node
            continue
        tail.next = node
        tail = node
    return head


def array_from_list(head):
    out = []
    current = head
    while current is not None:
        out.append(current.val)
        current = current.next
    return out


class TestAddTwoNums(unittest.TestCase):
    def assert_sum(self, l1, l2, expected):
        result = Solution().addTwoNumbers(list_from_array(l1), list_from_array(l2))
        self.assertEqual(array_from_list(result), expected)

    def test_example_1(self):
        self.assert_sum([2, 4, 3], [5, 6, 4], [7, 0, 8])

    def test_example_2(self):
        self.assert_sum([0], [0], [0])

    def test_example_3(self):
        self.assert_sum([9, 9, 9, 9, 9, 9, 9], [9, 9, 9, 9], [8, 9, 9, 9, 0, 0, 0, 1])


if __name__ == "__main__":
    unittest.main()
